From 10fdedbe2a5e02d34124e0937d79618da2199d15 Mon Sep 17 00:00:00 2001
From: vmallika <vmallika@redhat.com>
Date: Thu, 3 Dec 2015 14:55:36 +0530
Subject: [PATCH 172/178] glusterd/quota: quota-version conflict in export/import volinfo

This is a backport of http://review.gluster.org/#/c/12865/

When exporting/importing voinfo during handshake,
quota conf and quota xattr version were using same key
'quota-version' and updated wrong values when importing
quota version values.

> Change-Id: If939d6f5bc4851d4114963877be72dda21834f0f
> BUG: 1287996
> Signed-off-by: vmallika <vmallika@redhat.com>

Change-Id: Ic99d715bfe4f5b79d98f7c30303369f462128c8b
BUG: 1287980
Signed-off-by: vmallika <vmallika@redhat.com>
Reviewed-on: https://code.engineering.redhat.com/gerrit/63352
Reviewed-by: Atin Mukherjee <amukherj@redhat.com>
Tested-by: Atin Mukherjee <amukherj@redhat.com>
---
 tests/bugs/quota/bug-1287996.t              |   24 ++++++++++++++++++++++++
 xlators/mgmt/glusterd/src/glusterd-quota.c  |    6 +++---
 xlators/mgmt/glusterd/src/glusterd-store.c  |    4 ++--
 xlators/mgmt/glusterd/src/glusterd-utils.c  |    9 +++++----
 xlators/mgmt/glusterd/src/glusterd-volgen.c |    2 +-
 xlators/mgmt/glusterd/src/glusterd.h        |    2 +-
 6 files changed, 36 insertions(+), 11 deletions(-)
 create mode 100644 tests/bugs/quota/bug-1287996.t

diff --git a/tests/bugs/quota/bug-1287996.t b/tests/bugs/quota/bug-1287996.t
new file mode 100644
index 0000000..d75b7a2
--- /dev/null
+++ b/tests/bugs/quota/bug-1287996.t
@@ -0,0 +1,24 @@
+#!/bin/bash
+
+. $(dirname $0)/../../include.rc
+. $(dirname $0)/../../cluster.rc
+
+function check_peers {
+        $CLI_1 peer status | grep 'Peer in Cluster (Connected)' | wc -l
+}
+
+cleanup;
+
+TEST launch_cluster 2;
+
+TEST $CLI_1 volume create $V0 $H1:$B1/$V0
+TEST $CLI_1 volume start $V0
+TEST $CLI_1 volume quota $V0 enable
+
+TEST $CLI_1 peer probe $H2;
+EXPECT_WITHIN $PROBE_TIMEOUT 1 check_peers
+
+
+TEST $CLI_1 volume stop $V0
+TEST $CLI_1 volume delete $V0
+cleanup;
diff --git a/xlators/mgmt/glusterd/src/glusterd-quota.c b/xlators/mgmt/glusterd/src/glusterd-quota.c
index 1830f0e..6495aa4 100644
--- a/xlators/mgmt/glusterd/src/glusterd-quota.c
+++ b/xlators/mgmt/glusterd/src/glusterd-quota.c
@@ -1504,12 +1504,12 @@ glusterd_op_quota (dict_t *dict, char **op_errstr, dict_t *rsp_dict)
 
 
         if (GF_QUOTA_OPTION_TYPE_ENABLE == type)
-                volinfo->quota_version++;
+                volinfo->quota_xattr_version++;
         ret = glusterd_store_volinfo (volinfo,
                                       GLUSTERD_VOLINFO_VER_AC_INCREMENT);
         if (ret) {
                 if (GF_QUOTA_OPTION_TYPE_ENABLE == type)
-                        volinfo->quota_version--;
+                        volinfo->quota_xattr_version--;
                 goto out;
         }
 
@@ -1520,7 +1520,7 @@ glusterd_op_quota (dict_t *dict, char **op_errstr, dict_t *rsp_dict)
                                                   "volfiles");
                 if (GF_QUOTA_OPTION_TYPE_ENABLE == type) {
                         /* rollback volinfo */
-                        volinfo->quota_version--;
+                        volinfo->quota_xattr_version--;
                         ret = glusterd_store_volinfo (volinfo,
                                       GLUSTERD_VOLINFO_VER_AC_INCREMENT);
                 }
diff --git a/xlators/mgmt/glusterd/src/glusterd-store.c b/xlators/mgmt/glusterd/src/glusterd-store.c
index 83b6a96..f44623e 100644
--- a/xlators/mgmt/glusterd/src/glusterd-store.c
+++ b/xlators/mgmt/glusterd/src/glusterd-store.c
@@ -1012,7 +1012,7 @@ glusterd_volume_exclude_options_write (int fd, glusterd_volinfo_t *volinfo)
                         goto out;
         }
 
-        snprintf (buf, sizeof (buf), "%d", volinfo->quota_version);
+        snprintf (buf, sizeof (buf), "%d", volinfo->quota_xattr_version);
         ret = gf_store_save_value (fd, GLUSTERD_STORE_KEY_VOL_QUOTA_VERSION,
                                    buf);
         if (ret)
@@ -2679,7 +2679,7 @@ glusterd_store_update_volinfo (glusterd_volinfo_t *volinfo)
                         volinfo->tier_info.cold_type = atoi (value);
                 } else if (!strncmp (key, GLUSTERD_STORE_KEY_VOL_QUOTA_VERSION,
                             strlen (GLUSTERD_STORE_KEY_VOL_QUOTA_VERSION))) {
-                        volinfo->quota_version = atoi (value);
+                        volinfo->quota_xattr_version = atoi (value);
                 } else {
 
                         if (is_key_glusterd_hooks_friendly (key)) {
diff --git a/xlators/mgmt/glusterd/src/glusterd-utils.c b/xlators/mgmt/glusterd/src/glusterd-utils.c
index 5c547cd..a042207 100644
--- a/xlators/mgmt/glusterd/src/glusterd-utils.c
+++ b/xlators/mgmt/glusterd/src/glusterd-utils.c
@@ -2572,8 +2572,8 @@ glusterd_add_volume_to_dict (glusterd_volinfo_t *volinfo,
         ret = dict_set_int32 (dict, key, volinfo->caps);
 
         memset (key, 0, sizeof (key));
-        snprintf (key, sizeof (key), "%s%d.quota-version", prefix, count);
-        ret = dict_set_int32 (dict, key, volinfo->quota_version);
+        snprintf (key, sizeof (key), "%s%d.quota-xattr-version", prefix, count);
+        ret = dict_set_int32 (dict, key, volinfo->quota_xattr_version);
 out:
         GF_FREE (volume_id_str);
         GF_FREE (rebalance_id_str);
@@ -3655,9 +3655,10 @@ glusterd_import_volinfo (dict_t *peer_data, int count,
         ret = dict_get_int32 (peer_data, key, &new_volinfo->caps);
 
         memset (key, 0, sizeof (key));
-        snprintf (key, sizeof (key), "%s%d.quota_version", prefix, count);
+        snprintf (key, sizeof (key), "%s%d.quota-xattr-version", prefix, count);
         /*This is not present in older glusterfs versions, so ignore ret value*/
-        ret = dict_get_int32 (peer_data, key, &new_volinfo->quota_version);
+        ret = dict_get_int32 (peer_data, key,
+                              &new_volinfo->quota_xattr_version);
 
         ret = glusterd_import_bricks (peer_data, count, new_volinfo, prefix);
         if (ret)
diff --git a/xlators/mgmt/glusterd/src/glusterd-volgen.c b/xlators/mgmt/glusterd/src/glusterd-volgen.c
index 0a8f4c5..a200623 100644
--- a/xlators/mgmt/glusterd/src/glusterd-volgen.c
+++ b/xlators/mgmt/glusterd/src/glusterd-volgen.c
@@ -1902,7 +1902,7 @@ brick_graph_add_marker (volgen_graph_t *graph, glusterd_volinfo_t *volinfo,
         if (ret)
                 goto out;
 
-        snprintf (buf, sizeof (buf), "%d", volinfo->quota_version);
+        snprintf (buf, sizeof (buf), "%d", volinfo->quota_xattr_version);
         ret = xlator_set_option (xl, "quota-version", buf);
         if (ret)
                 goto out;
diff --git a/xlators/mgmt/glusterd/src/glusterd.h b/xlators/mgmt/glusterd/src/glusterd.h
index a51320f..1e573f8 100644
--- a/xlators/mgmt/glusterd/src/glusterd.h
+++ b/xlators/mgmt/glusterd/src/glusterd.h
@@ -433,7 +433,7 @@ struct glusterd_volinfo_ {
         gd_quorum_status_t        quorum_status;
 
         glusterd_snapdsvc_t       snapd;
-        int32_t                   quota_version;
+        int32_t                   quota_xattr_version;
 };
 
 typedef enum gd_snap_status_ {
-- 
1.7.1

